app/components/ledger/ledger-connect-instructions.tsx

```diff
@@ -1,35 +1,49 @@
 
-import { LedgerConnectStep } from '../../pages/onboarding';
 import { LedgerStepText } from './ledger-step-text';
+import { LedgerConnectStep } from '@hooks/use-prepare-ledger';
 
 interface LedgerConnectInstructions {
   step: LedgerConnectStep;
   action: string;
+  isLocked: boolean;
 }
 
-export const LedgerConnectInstructions: FC<LedgerConnectInstructions> = ({ step, action }) => {
+export const LedgerConnectInstructions: FC<LedgerConnectInstructions> = props => {
+  const { step, action, isLocked } = props;
+
   const hasConnected = (step: LedgerConnectStep) => step > LedgerConnectStep.Disconnected;
   const hasOpenedApp = (step: LedgerConnectStep) => step > LedgerConnectStep.ConnectedAppClosed;
-  const hasAddress = (step: LedgerConnectStep) => step === LedgerConnectStep.HasAddress;
+  const hasAddress = (step: LedgerConnectStep) => step === LedgerConnectStep.ActionComplete;
 
   return (
-    <Box border="1px solid #F0F0F5" mt="extra-loose" borderRadius="8px">
-      <Flex height="56px" alignItems="center" px="extra-loose" borderBottom="1px solid #F0F0F5">
-        <LedgerStepText step={LedgerConnectStep.Disconnected}>Connect your Ledger</LedgerStepText>
-        {hasConnected(step) && <CheckmarkCircleIcon color="blue" size="16px" ml="tight" />}
-      </Flex>
-      <Flex height="56px" alignItems="center" px="extra-loose" borderBottom="1px solid #F0F0F5">
-        <LedgerStepText step={LedgerConnectStep.ConnectedAppClosed}>
-          Open the Stacks app
-        </LedgerStepText>
-        {hasOpenedApp(step) && <CheckmarkCircleIcon color="blue" size="16px" ml="tight" />}
-      </Flex>
-      <Flex height="56px" alignItems="center" px="extra-loose">
-        <LedgerStepText step={LedgerConnectStep.ConnectedAppOpen}>{action}</LedgerStepText>
-        {hasAddress(step) && <CheckmarkCircleIcon color="blue" size="16px" ml="tight" />}
-      </Flex>
-    </Box>
+    <>
+      <Box border="1px solid #F0F0F5" mt="extra-loose" borderRadius="8px">
+        <Flex height="56px" alignItems="center" px="extra-loose" borderBottom="1px solid #F0F0F5">
+          <LedgerStepText step={LedgerConnectStep.Disconnected}>Connect your Ledger</LedgerStepText>
+          {hasConnected(step) && <CheckmarkCircleIcon color="blue" size="16px" ml="tight" />}
+        </Flex>
+        <Flex height="56px" alignItems="center" px="extra-loose" borderBottom="1px solid #F0F0F5">
+          <LedgerStepText step={LedgerConnectStep.ConnectedAppClosed}>
+            Open the Stacks app
+          </LedgerStepText>
+          {hasOpenedApp(step) && <CheckmarkCircleIcon color="blue" size="16px" ml="tight" />}
+        </Flex>
+        <Flex height="56px" alignItems="center" px="extra-loose">
+          <LedgerStepText step={LedgerConnectStep.ConnectedAppOpen}>{action}</LedgerStepText>
+          {hasAddress(step) && <CheckmarkCircleIcon color="blue" size="16px" ml="tight" />}
+        </Flex>
+      </Box>
```

Comment:
Purely a "ui-engineering" nit. I would prefer if this component were to be abstracted out, there is a lot of duplication here. eg:

```tsx
import React, { FC } from 'react';
import { Box, Flex, Text, CheckmarkCircleIcon, EncryptionIcon, FlexProps } from '@blockstack/ui';

import { LedgerStepText } from './ledger-step-text';
import { LedgerConnectStep } from '@hooks/use-prepare-ledger';

interface LedgerConnectInstructions {
  step: LedgerConnectStep;
  action: string;
  isLocked: boolean;
}

interface ListItemProps extends FlexProps {
  isComplete: boolean;
  step: LedgerConnectStep;
  label: string;
  isLast?: boolean;
}

const ListItem: React.FC<ListItemProps> = ({ isComplete, step, label, isLast, ...rest }) => {
  return (
    <Flex
      height="56px"
      alignItems="center"
      px="extra-loose"
      borderBottom={!isLast ? '1px solid #F0F0F5' : 'unset'}
      {...rest}
    >
      <LedgerStepText step={LedgerConnectStep.Disconnected}>{label}</LedgerStepText>
      {isComplete && <CheckmarkCircleIcon color="blue" size="16px" ml="tight" />}
    </Flex>
  );
};

export const LedgerConnectInstructions: FC<LedgerConnectInstructions> = props => {
  const { step, action, isLocked } = props;

  const hasConnected = (step: LedgerConnectStep) => step > LedgerConnectStep.Disconnected;
  const hasOpenedApp = (step: LedgerConnectStep) => step > LedgerConnectStep.ConnectedAppClosed;
  const hasAddress = (step: LedgerConnectStep) => step === LedgerConnectStep.ActionComplete;

  const items = [
    {
      step: LedgerConnectStep.Disconnected,
      label: 'Connect your ledger',
      isComplete: hasConnected(step),
    },
    {
      step: LedgerConnectStep.ConnectedAppClosed,
      label: 'Open the Stacks app',
      isComplete: hasOpenedApp(step),
    },
    {
      step: LedgerConnectStep.ConnectedAppOpen,
      label: action,
      isComplete: hasAddress(step),
    },
  ];

  return (
    <>
      <Box border="1px solid #F0F0F5" mt="extra-loose" borderRadius="8px">
        {items.map((item, index, arr) => (
          <ListItem
            isComplete={item.isComplete}
            step={item.step}
            label={item.label}
            key={item.step}
            isLast={index === arr.length - 1}
          />
        ))}
      </Box>
      {isLocked && (
        <Flex mt="base">
          <EncryptionIcon size="12px" mt="2px" color="ink.600" />
          <Text textStyle="caption" ml="extra-tight" color="ink.600">
            Your Ledger device is locked. To continue, press a button your and device and enter your
            PIN code.
          </Text>
        </Flex>
      )}
    </>
  );
};
```

Reply:
Yep, agree. Tidied this up.

Though, I've not gone as far as looping an array. This abstraction makes it harder to accommodate change if, say, the styling of 1 changes.

```tsx
        <LedgerConnectStepRow isComplete={hasConnected(step)}>
          <LedgerStepText step={LedgerConnectStep.Disconnected}>Connect your Ledger</LedgerStepText>
        </LedgerConnectStepRow>

        <LedgerConnectStepRow isComplete={hasOpenedApp(step)}>
          <LedgerStepText step={LedgerConnectStep.ConnectedAppClosed}>
            Open the Stacks app
          </LedgerStepText>
        </LedgerConnectStepRow>

        <LedgerConnectStepRow isComplete={hasAddress(step)} isLast>
          <LedgerStepText step={LedgerConnectStep.ConnectedAppOpen}>{action}</LedgerStepText>
        </LedgerConnectStepRow>
```

Reply:
and moved these functions out of the component 👍🏼